go/cmd/relations/upstream.go

```diff
@@ -0,0 +1,272 @@
+			if _, ok := visited[upstream]; !ok && !slices.Contains(toVisit, upstream) {
+				toVisit = append(toVisit, upstream)
+			}
+		}
+	}
+
+	// Ensure there are no duplicate entries where transitive vulns appear
+	for k, v := range upstreamMap {
+		if k == targetUpstreamGroup.VulnID {
+			continue
+		}
+		newGroup := make([]string, 0, len(upstreamMap[targetUpstreamGroup.VulnID]))
+		for _, upstream := range upstreamMap[targetUpstreamGroup.VulnID] {
+			if !slices.Contains(v, upstream) {
+				newGroup = append(newGroup, upstream)
+			}
+		}
+		upstreamMap[targetUpstreamGroup.VulnID] = newGroup
+	}
+
+	// Update the datastore entry if hierarchy has changed
+	if len(upstreamMap) == 0 {
+		return nil
+	}
+	// Sort the upstreams to ensure consistent ordering
+	for _, v := range upstreamMap {
+		slices.Sort(v)
+	}
+	upstreamJSON, err := json.Marshal(upstreamMap)
+	if err != nil {
+		return err
+	}
+	if bytes.Equal(upstreamJSON, targetUpstreamGroup.UpstreamHierarchy) {
+		return nil
+	}
+	targetUpstreamGroup.UpstreamHierarchy = upstreamJSON
+	_, err = cl.Put(ctx, targetUpstreamGroup.Key, targetUpstreamGroup)
+
+	return err
+}
+
+func ComputeUpstreamGroups(ctx context.Context, cl *datastore.Client, ch chan<- Update) error {
+	// Query for all vulns that have upstreams.
+	var updatedGroups []*models.UpstreamGroup
+	logger.Info("Retrieving vulns for upstream computation...")
+	query := datastore.NewQuery("Vulnerability").FilterField("upstream_raw", ">", "")
+	it := cl.Run(ctx, query)
+
+	vulns := make(map[string][]string)
+	for {
+		var vuln models.Vulnerability
+		_, err := it.Next(&vuln)
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("failed to iterate vulnerabilities: %w", err)
+		}
+		upstream := slices.Clone(vuln.UpstreamRaw)
+		slices.Sort(upstream)
+		upstream = slices.Compact(upstream)
+		vulns[vuln.Key.Name] = upstream
+	}
+	logger.Info("Vulns successfully retrieved", slog.Int("count", len(vulns)))
+
+	logger.Info("Retrieving upstream groups...")
+	query = datastore.NewQuery("UpstreamGroup")
+	it = cl.Run(ctx, query)
+	upstreamGroups := make(map[string]*models.UpstreamGroup)
+	for {
+		var group models.UpstreamGroup
+		_, err := it.Next(&group)
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("failed to iterate upstream groups: %w", err)
+		}
+		upstreamGroups[group.VulnID] = &group
+	}
+	logger.Info("Upstream groups successfully retrieved", slog.Int("count", len(upstreamGroups)))
+
+	for vulnID, upstreams := range vulns {
+		// Get the specific upstream group ID
+		group, exists := upstreamGroups[vulnID]
+		// Recompute the transitive upstreams and compare with the existing group
+		upstreamIDs := computeUpstream(upstreams, vulns)
+		if exists {
+			if slices.Equal(upstreamIDs, group.UpstreamIDs) {
+				continue
+			}
+			// Update the existing UpstreamGroup
+			var err error
+			group, err = updateUpstreamGroup(ctx, cl, group, upstreamIDs, ch)
```

Comment:
Not sure what to do about this, but it is a bit weird that this function both returns a new value, and also does the datastore updates.